### PR TITLE
added login error box 

### PR DESCRIPTION
**Pull request by lenagrech**
the original ticket, but i set up everything from scratch because of a git workflow issue i had ( now fixed) 

now i should be based on the latest master as well

### CODE REVIEW

**Comment by orronai** [comment]
Next time when you open a new PR, doesn't matter if it's connected to another PR:
1. Commit like this format: https://www.conventionalcommits.org/en/v1.0.0/
2. Code modification check list before pushing: https://github.com/PythonFreeCourse/lms#code-modification-check-list

File: lms/lmsweb/views.py [hunk 1]
```diff
@@ -100,7 +100,8 @@ def login(login_error: str = None):
             login_error = 'Invalid username or password'
             return redirect(
                 url_for(
-                    'login', **{'next': next_page, 'login_error': login_error}),
+                    'login', **{'next': next_page,
+                                'login_error': login_error}),
```
Comment (orronai) [inline lms/lmsweb/views.py:104]: I'd prefer:
`return redirect(url_for(
    `your parameters`,
    `other lines of parameters if needed`,
))

Comment (lenagrech) [reply lms/lmsweb/views.py:104]: thanks, I appreciate the suggestion. I think Yam told me last time he prefers some other format so I'll wait for his input on this one

File: lms/lmsweb/views.py [hunk 2]
```diff
@@ -81,22 +81,30 @@ def ratelimit_handler(e):
     f'{LIMITS_PER_MINUTE}/minute;{LIMITS_PER_HOUR}/hour',
     deduct_when=lambda response: response.status_code != 200,
 )
-def login():
+def login(login_error: str = None):
```
Comment (yammesicka) [inline lms/lmsweb/views.py:84]: If it gets `None` as default value, it must be annotated as `Optional[str]`

File: lms/lmsweb/views.py (hunk 1 above)
Comment (yammesicka) [reply lms/lmsweb/views.py:104]: I think the best solution for this one is:

```python
error_details = {'next': next_page, 'login_error': login_error}
return redirect(url_for('login', **error_details))

File: lms/lmsweb/views.py [hunk 3]
```diff
@@ -81,22 +81,30 @@ def ratelimit_handler(e):
...
+            return get_next_url(next_page)
+        elif user is None or user.is_password_valid(password) is False:
+            login_error = 'Invalid username or password'
+            return redirect(
+                url_for(
+                    'login', **{'next': next_page,
+                                'login_error': login_error}),
+            )
+
+    return render_template('login.html', login_error=login_error)
```
Comment (yammesicka) [inline lms/lmsweb/views.py:107]: This code is very sensitive from the security perspective, hence I want it to have test-coverage of 100%. Can you please create login test file that check all the possible scenarios?

@orronai has already written some tests in this regards, maybe you get to the goal of 100% coverage together :)

**Review by yammesicka — CHANGES_REQUESTED** [review]

File: lms/lmsweb/views.py [hunk 4]
```diff
@@ -81,22 +81,27 @@ def ratelimit_handler(e):
...
-    elif user is not None:
-        return redirect(url_for('login', **{'next': next_page}))
-
-    return render_template('login.html')
+    if request.method == 'POST':
+        if user is not None and user.is_password_valid(password):
+            login_user(user)
+            return get_next_url(next_page)
+        elif user is None or user.is_password_valid(password) is False:
+            login_error = 'Invalid username or passwordd'
```
Comment (orronai) [inline lms/lmsweb/views.py:100]: Password with one `d`

File: lms/lmsweb/views.py (hunk 3 above)
Comment (lenagrech) [reply lms/lmsweb/views.py:104]: i added tests for login errors and success

**Comment by yammesicka** [comment]
Awesome job, @lenagrech!

**Pull request merged**